pip/pip-385.md

```diff
@@ -0,0 +1,398 @@
+            Client-->>Server: ThrottleProducerReceipt(reqId)
+            note right of Client: client pauses till specified time
+            Note over Client, Server: After some time
+            Client->>Server: Send(3, message3)
+            Client->>Server: Send(4, message4)
+
+        else no response
+            Activate Server
+            note right of Server: TCP read paused
+            Client-xServer: Send(3, message3)
+            Note over Client, Server: After some time
+            opt topic/connection unthrottled
+                note right of Server: TCP read resumed
+                deactivate Server
+                Server-->>Server: read message 3
+                Client->>Server: Send(4, message4)
+            end
+        end
+    end
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Server-->>Client: SendReceipt(4,msgId4)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Java client changes
+
+* Client will now have logic to understand the `ThrottleProducer` command and take relevant action of blocking further
+  messages for the relevant topic. It will then respond back with `ThrottleProducerReceipt` command.
+    * Client will resume message sending after the specified time in the `ThrottleProducer` command's data.
+    * This interval of no messages will be noted as "being throttled"
+    * Within this duration, another `ThrottleProducer` command from server may come.
+* Producer will record new OTel metric indicating which topic was throttled and the reason.
+* In case a message fails due to timeout and there was a throttled command from server for the owning topic, client will
+  instead throw a rate limit exception instead of timeout exception.
+
+# Detailed Design
+
+## High-level Implementation Details
+
+### Broker Changes
+
+* For calls arising from `PublishRateLimiterImpl` class, add logic in `ServerCnxThrottleTracker.java` to send the
+  command to client and wait for response for the max configured duration before calling `changeAutoRead`. It checks for
+  feature availability first.
+* For calls arising from `ServerCnxThrottleTracker::changeThrottlingFlag`, we send the command async (if feature
+  supported) without worrying about response and then call `changeAutoRead`.
+* capture `AbstractTopic::getTotalPublishRateLimitCounter` per publish rate limit counter and add relevant attribute in
+  the rate limit metric
+
+### Determining the throttling duration for clients
+
+Broadly, there are 2 different categories for rate limiting - (a) quota exceeding at a topic/resource group level OR (b)
+connection level breaches leading to broker's resource crunch.
+
+In case (a), as per the current token bucket based quota tracking introduced in PIP-322, we (lazily) resume channel
+reads on the next second mark. i.e. the throttling is only applied at max for 1 second.
```

Comment:
After PIP-322, there's no longer any max 1 second throttling. That was the case before PIP-322 since the tokens in the token bucket couldn't get negative. After PIP-322, it will count all tokens and throttle until there's tokens available. You can find the implementation in org.apache.pulsar.broker.qos.AsyncTokenBucket#calculateThrottlingDuration: https://github.com/apache/pulsar/blob/c4cff0ab3dac169c0a1418ef2f63f61604f6278e/pulsar-broker/src/main/java/org/apache/pulsar/broker/qos/AsyncTokenBucket.java#L288-L306
Used in PublishRateLimiterImpl to calculate the throttling duration:
https://github.com/apache/pulsar/blob/c4cff0ab3dac169c0a1418ef2f63f61604f6278e/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiterImpl.java#L112-L124

Reply:
@lhotari but the fact that new tokens are added (lazily) every second to the bucket effectively limit the throttling to 1 second, right? the producers may get throttled right away the very next nano second again ..

Reply:
> @lhotari but the fact that new tokens are added (lazily) every second to the bucket effectively limit the throttling to 1 second, right? the producers may get throttled right away the very next nano second again ..

No, after PIP-322, tokens will be calculated based on the elapsed time and the configured rate of adding tokens. The rate period of 1 second is only impacting how the rate of adding tokens is expressed. It's possible that some of the current documentation is outdated and hasn't been updated to capture this detail.
A different matter is that the maximum number of tokens in the bucket is limited by the bucket size. It's currently set to the configured rate since in Pulsar, we have a single number for configuring publishing rate. The DTO class is https://github.com/apache/pulsar/blob/master/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/PublishRate.java .
After PIP-322, there's no resetting of the tokens and the number of consumed tokens will often become negative. That is necessary in an asynchronous token bucket implementation like there is in Pulsar. When tokens are consumed, the operation is never blocked when that happens. The throttling will happen when tokens become negative and the throttling period will last until time has elapsed until there's a minimum amount of tokens in the bucket again. By default, the minimum amount of tokens is 16 milliseconds of worth tokens with the configured rate. Let's say if the rate is 1000 / second and tokens have gone 500 tokens negative, the throttle duration will be 516 milliseconds. This is how the current rate limiting works in Pulsar after PIP-322.

When PIP-322 started, we had discussions around having configurable options for allowing bursting. That's all possible with the current foundation laid with PIP-322 changes. The simplest form of bursting configuration would be to have separate configuration options for the bucket size so that it's not dependent on the configuration of the rate of adding tokens as it is currently. There could also be variable rates for adding tokens as we have discussed in the past, but that would require additional work to implement it. Adding basic bursting support would be something to consider before adding more advanced features since bursting support is already technically implemented with the token bucket size and adding bursting support is mainly about adding the configuration options and documentation for that.

Reply:
I was busy with other production issue during the PIP-322 time and it was merged (impl included) within a week or two.. but overall, my understanding was that in practice, we are sticking to the strict rate limiter design and only the under the hood implementation is being changed to be more efficient and future proof?

But now the way you are explaining that tokens could be -500 - seems like we have gone back to the non-existent ex-default poller based rate limiter that was there prior to 322? What is preventing a topic to burst and reach -10000 tokens? Sure, it would take longer for the bucket to come back to 0 and thus, the produce to that topic would be blocked for longer time than the next second mark, but the damage would have already been done with respect to noisy neighbor (at the broker level) due to exhausting other physical resources like disk bandwidth etc..

I will have to read the code in detail to understand this better.. but please correct me if the above understanding is wrong.

Having said that, if my understanding is correct - isn't this a much bigger challenge than the concern you have raised in your comment regarding calculating RTT?

Reply:
@lhotari  - Can you please provide some insights here? 

> But now the way you are explaining that tokens could be -500 - seems like we have gone back to the non-existent ex-default poller based rate limiter that was there prior to 322? What is preventing a topic to burst and reach -10000 tokens? Sure, it would take longer for the bucket to come back to 0 and thus, the produce to that topic would be blocked for longer time than the next second mark, but the damage would have already been done with respect to noisy neighbor (at the broker level) due to exhausting other physical resources like disk bandwidth etc..

Reply:
In Pulsar, the rate limiting is non-blocking and it doesn't drop traffic. For Netty, it's not a feasible approach to block in the network layer. This is one of the reasons for the non-blocking approach. The implications of this is that when a client sends (publishes) messages and it arrives to the broker, it will always be handled as fast as possible. This is the reason why it must be allowed to have as many negative tokens as has been published. The rate will smoothen over time since the throttling delay will last until there are a minimal amount of tokens available. 

> but overall, my understanding was that in practice, we are sticking to the strict rate limiter design and only the under the hood implementation is being changed to be more efficient and future proof?

What' your definition of "strict rate limiter" design? 

> But now the way you are explaining that tokens could be -500 - seems like we have gone back to the non-existent ex-default poller based rate limiter that was there prior to 322? What is preventing a topic to burst and reach -10000 tokens? 

I hope that the explanation of the non-blocking nature of Pulsar's rate limiting in the first paragraph of this comment explains why tokens can go to any negative value that is necessary to handle the already in progress messages until throttling can be applied to prevent new messages to enter the broker.

> Sure, it would take longer for the bucket to come back to 0 and thus, the produce to that topic would be blocked for longer time than the next second mark, but the damage would have already been done with respect to noisy neighbor (at the broker level) due to exhausting other physical resources like disk bandwidth etc..

In PIP-322, there's no "next second mark". I hope that the explanation of the non-blocking and non-dropping nature of Pulsar rate limiting explains that.

> Having said that, if my understanding is correct - isn't this a much bigger challenge than the concern you have raised in your comment regarding calculating RTT?

What challenge are you referring to?

Reply:
> In Pulsar, the rate limiting is non-blocking and it doesn't drop traffic. For Netty, it's not a feasible approach to block in the network layer. This is one of the reasons for the non-blocking approach. The implications of this is that when a client sends (publishes) messages and it arrives to the broker, it will always be handled as fast as possible. This is the reason why it must be allowed to have as many negative tokens as has been published. The rate will smoothen over time since the throttling delay will last until there are a minimal amount of tokens available.

This wasn't the case before PIP-322.. Specifically, it was mentioned that overall user facing behavior isn't changing. It was mentioned multiple times:

https://github.com/apache/pulsar/pull/21680/files#diff-988544056ef2793c7360d39cdeb7f020416985d08bbc0072fee1a4aebf18246dR29

https://github.com/apache/pulsar/pull/21680/files#diff-988544056ef2793c7360d39cdeb7f020416985d08bbc0072fee1a4aebf18246dR179

https://github.com/apache/pulsar/pull/21680/files#diff-988544056ef2793c7360d39cdeb7f020416985d08bbc0072fee1a4aebf18246dR369

If I look at it from a user perspective, to me, this has now regressed from the way it was pre PIP-322. Let's take a look:

Before PIP-322:

* Default poller based throttling - we can define the polling frequency and bad producers can burst as much as they want between 2 polling checks, once every second.
* Precise poller based throttling - very strict rate limiter, doesn't allow a single byte/qps over the value. In theory, has locking contention, but in practice (I have personally done numerous different workload tests) has batter CPU performance than the default poller based throttling.

After PIP-322:

* Async token based throttling - bad producers can burst as much as they want between an unconfigurable time window. They cannot do this every second, but depends on how much they burst.

So, post PIP-322 there is no strict rate limiter, but possibly a some what better version of the default rate limiter pre-PIP-322 assuming that hardware doesn't degrade.

> What' your definition of "strict rate limiter" design?

I think its very straight forward - a rate limiter where "infinite" or even significant throughput above the limit is not allowed and is blocked.

> I hope that the explanation of the non-blocking nature of Pulsar's rate limiting in the first paragraph of this comment explains why tokens can go to any negative value that is necessary to handle the already in progress messages until throttling can be applied to prevent new messages to enter the broker.

Sure, in theory, it sounds good. But this is a MAJOR concern in practice. There can be bad actors leading to sudden spike in traffic. Sure, they will be blocked for relevant time, but when they come back again, they will spike as much as they want once again.

imagine this case - a topic with 1000 qps rate limit. So essentially 1000 tokens gets added every second spread over the 16ms time frame you've mentioned.
A bad actor can lead to a burst of 10,000 - then it gets throttled for 10 seconds, then it can again do 10,000.

At a macro level, it may seem that throttling is working very well and things are getting smoothed out, but that's not how hardware works. That single 10,000 burst would have added heavy toll to the underlying disk or network leading to noisy neighbor impact on all other topics present on that broker. How is this being contained?

> What challenge are you referring to?

the non existence of RTT in this proposal is practically a non-concern if the tokens can go to any humongous negative value. That's what I am referring to.

Reply:
> This wasn't the case before PIP-322.. Specifically, it was mentioned that overall user facing behavior isn't changing.

> If I look at it from a user perspective, to me, this has now regressed from the way it was pre PIP-322. Let's take a look:

From the user's perspective, PIP-322 is a significant improvement. There aren't any regressions or breaking changes in behavior. If you test the PIP-322 solution, you will notice that it behaves in very reasonable ways and provide good results. 

> Async token based throttling - bad producers can burst as much as they want between an unconfigurable time window. They cannot do this every second, but depends on how much they burst.

This isn't the case.
Pulsar has never dropped or rejected traffic when a producer goes over the limit. Before PIP-322, the behavior was that a bad producer can burst as much as they want. That has been fixed in PIP-322. 
If there's a burst and the tokens go negative, the TCP throttling will stop reading more messages from the network layer (Netty) and it will throttle until there are available tokens available. Due to the use of Netty's [io.netty.handler.flow.FlowControlHandler in the channel pipeline](https://github.com/apache/pulsar/blob/895e96853e56b7d9c10c6bd0f3bd6105b664eb14/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java#L101), Pulsar can accept messages one-by-one when it's throttling. This is how the tokens wouldn't go infinitely negative even when there's a large spike. FlowControlHandler has been in use before too so things haven't change there.
In Pulsar, we don't have a way to configure the token bucket size separately from the rate at the moment. The token bucket will fill up in 1 second and that's why spikes are limited to that 1 second worth of tokens.

> I think its very straight forward - a rate limiter where "infinite" or even significant throughput above the limit is not allowed and is blocked.

As mentioned above, there's no infinite rate limiter in PIP-322 implementation. The problem you are describing doesn't exist in PIP-322. 

> Sure, in theory, it sounds good. But this is a MAJOR concern in practice. There can be bad actors leading to sudden spike in traffic. Sure, they will be blocked for relevant time, but when they come back again, they will spike as much as they want once again.
> 
> imagine this case - a topic with 1000 qps rate limit. So essentially 1000 tokens gets added every second spread over the 16ms time frame you've mentioned.
> A bad actor can lead to a burst of 10,000 - then it gets throttled for 10 seconds, then it can again do 10,000.

This problem doesn't exist in the PIP-322 implementation. Messages (== batch messages, a.k.a "entries") are handled one-by-one with FlowControlHandler. The maximum amount it can go over the limit is by one entry.
If the configured rate is very low, 1 batch message can cause the tokens to go negative. That's a correct way to handle rate limiting in Pulsar.

> At a macro level, it may seem that throttling is working very well and things are getting smoothed out, but that's not how hardware works. That single 10,000 burst would have added heavy toll to the underlying disk or network leading to noisy neighbor impact on all other topics present on that broker. How is this being contained?

This is not a problem since the PIP-322 rate limiting is really a "strict" rate limiting implementation. I hope the above explanations clarify this.

> the non existence of RTT in this proposal is practically a non-concern if the tokens can go to any humongous negative value. That's what I am referring to.

tokens don't go to humongous negative values as explained above.

Reply:
> This isn't the case. Pulsar has never dropped or rejected traffic when a producer goes over the limit. Before PIP-322, the behavior was that a bad producer can burst as much as they want.

Not if the broker was configured to use PreciseRateLimieter. Moreover, even as per the default configurations, a bad producer could only burst for upto first 10ms of a every second.

> That has been fixed in PIP-322. If there's a burst and the tokens go negative, the TCP throttling will stop reading more messages from the network layer (Netty) and it will throttle until there are available tokens available. Due to the use of Netty's [io.netty.handler.flow.FlowControlHandler in the channel pipeline](https://github.com/apache/pulsar/blob/895e96853e56b7d9c10c6bd0f3bd6105b664eb14/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarChannelInitializer.java#L101), Pulsar can accept messages one-by-one when it's throttling. This is how the tokens wouldn't go infinitely negative even when there's a large spike.  FlowControlHandler has been in use before too so things haven't change there. In Pulsar, we don't have a way to configure the token bucket size separately from the rate at the moment. The token bucket will fill up in 1 second and that's why spikes are limited to that 1 second worth of tokens.

I am not talking about the case when it's already throttling. I am talking about the case during the 16ms (Default resolution time, and as far as I see, non-configurable by pulsar users), the token check takes a fast path as `shouldUpdateTokensImmediately ` returns false and thus tokens are just added into a `pendingConsumedTokens` variable instead of actually checking if there are enough tokens left in the bucket.

ref: https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/qos/AsyncTokenBucket.java#L153

Please let me know if my read on this code piece is incorrect, but as I understand, multiple producers can send multiple entries during the `resolutionNanos` time frame and they all will be allowed through. This can even go beyond the  1 second worth of tokens cap that you are referring to.

> > imagine this case - a topic with 1000 qps rate limit. So essentially 1000 tokens gets added every second spread over the 16ms time frame you've mentioned.
> > A bad actor can lead to a burst of 10,000 - then it gets throttled for 10 seconds, then it can again do 10,000.
> 
> This problem doesn't exist in the PIP-322 implementation. Messages (== batch messages, a.k.a "entries") are handled one-by-one with FlowControlHandler. The maximum amount it can go over the limit is by one entry. If the configured rate is very low, 1 batch message can cause the tokens to go negative. That's a correct way to handle rate limiting in Pulsar.

Can you point me to the code piece which enforces this? My read on the code and the java doc is different. I followed the code, and I see no such check that only allows 1 batch message to make the tokens go negative. As I read the code (referenced above), any number of batches can be produced within the `resolutionNanos` time duration without actually reducing the tokens.